### DISCUSSION TITLE
Libraries and runtime updates in .NET 9 Preview 4

### DISCUSSION THREAD

**Discussion by richlander** [dotnet/runtime#102504] [Announcements]
Here's a summary of what's new in .NET Libraries in this preview release:

- [New Tensor<T> type](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview4/libraries.md#new-tensort-type)
- [Tokenizer library enhancements](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview4/libraries.md#tokenizer-library-enhancements)
- [OpenTelemetry: Make activity linking more flexible](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview4/libraries.md#opentelemetry-make-activity-linking-more-flexible)
- [PDB support for System.Reflection.Emit.PersistedAssemblyBuilder](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview4/libraries.md#pdb-support-added-for-systemreflectionemitpersistedassemblybuilder)

Here's a summary of what's new in the .NET Runtime in this preview release:

- [`UnsafeAccessorAttribute` supports generic parameters](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview4/runtime.md#unsafeaccessorattribute-supports-generic-parameters)
- [Feature switches with trimming support](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview4/runtime.md#feature-switches-with-trimming-support)

Release notes:

- [Libraries release notes](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview4/libraries.md)
- [Runtime release notes](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview4/runtime.md)

.NET 9 Preview 4:

- [Discussion](https://aka.ms/dotnet/9/preview4)
- [Release notes](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview4/README.md)
- [What's new in .NET 9](https://learn.microsoft.com/dotnet/core/whats-new/dotnet-9/overview)

**Comment by KalleOlaviNiemitalo** [comment]
"Tokenizer library enhancements" seems to refer to <https://github.com/dotnet/machinelearning/tree/v4.0.0-preview1/src/Microsoft.ML.Tokenizers>. Please consider linking from the release notes to <https://dot.net/ml> and <https://learn.microsoft.com/dotnet/api/microsoft.ml.tokenizers.tokenizer?view=ml-dotnet-preview>, to clarify what this is about; but the API changes after version 0.21.1 of the Microsoft.ML.Tokenizers package are not documented there yet.

**Reply by jamesmontemagno** [reply]
@luisquintanilla

**Reply by luisquintanilla** [reply]
Thanks for the feedback @KalleOlaviNiemitalo. We'll work to get those docs updated.

**Reply by KalleOlaviNiemitalo** [reply]
The example in <https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview4/libraries.md#tokenizer-library-enhancements> isn't valid either…

```csharp
Span<char> textSpan = "Hello World".AsSpan();
```

AsSpan here returns ReadOnlySpan\<char\> rather than Span\<char\>.

**Reply by luisquintanilla** [reply]
cc: @tarekgh

**Reply by tarekgh** [reply]
I have created the PR https://github.com/dotnet/core/pull/9327. Please let me know if you have any more comments. Thanks!

**Reply by KeterSCP** [reply]
https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview4/libraries.md has invalid link to `New Tensor<T> type`.  The link is `#new-tensor-type`, but it should be `#new-tensort-type` (notice the additional **t**)

**Reply by richlander** [reply]
Thanks for reporting. 

https://github.com/dotnet/core/pull/9330

**Reply by richlander** [reply]
FYI: The community is free to make PRs on release notes.

**Comment by ReneeGA2020** [comment]
It looks like **dotnet workload install wasi-experimental** always fail on Downloading Aspire.Hosting.Sdk.Msi.x64. I cannot install this workload in preview 4.

**Reply by richlander** [reply]
@lewing

**Reply by ReneeGA2020** [reply]
It worked now.

[Edited]
ok, not.

I reinstalled preview 4,and tried it again, it failed again.
Error NU3004, this package is unsigned: Aspire.Hosting.Sdk.Msi.x64.9.0.0-preview.2.24163.9

**Reply by richlander** [reply]
@baronfel

**Reply by ReneeGA2020** [reply]
Ok, I figured it out. In an earlier day, I added dotnet 9 package source into the global package manager.

 https://pkgs.dev.azure.com/dnceng/public/_packaging/dotnet9/nuget/v3/index.json

It looks like this package source is no longer needed for preview 4, and was causing conflicts.

I'm now able to download and install dotnet 9 workloads correctly after deleted this package source.

**Comment by PaskalSunari** [comment]
### I have installed preview 4 created the project with target framework net 9.0..But the build error is thrown while building the project.. does anybody has the solution for this error..

Severity	Code	Description	Project	File	Line	Suppression State
Error (active)	NETSDK1045	The current .NET SDK does not support targeting .NET 9.0.  Either target .NET 8.0 or lower, or use a version of the .NET SDK that supports .NET 9.0. Download the .NET SDK from https://aka.ms/dotnet/download	MyFirstDotNet9WebApp	C:\Program Files\dotnet\sdk\8.0.300\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.TargetFrameworkInference.targets	166	

>

**Reply by huoyaoyuan** [reply]
To use a preview version of .NET SDK in Visual Studio, you should do either of following:
- Enable "Use previews of the .NET SDK" in Visual Studio Options->Environment->Preview Features
- Use a [global.json](https://learn.microsoft.com/dotnet/core/tools/global-json) that specifies the preview SDK version.
